### PR TITLE
Updated dict .keys() and .values() indexed calls for py3

### PR DESCRIPTION
**Pull request by raddessi**
### What does this PR do?
Update all indexed dict .keys() and .values() calls I could find to be compatible with python 3, since both `dict_keys` and `dict_values` objects are not subscriptable in python 3.

### What issues does this PR fix or reference?
I found this from napalm beacons not working when running in py3.

### Previous Behavior
Napalm beacons would fail to run on py3.

### New Behavior
Napalm beacons work as expected

### Tests written?
No

### Commits signed with GPG?
No

### CODE REVIEW

**Review by s0undt3ch — COMMENTED** [review]
Hmm, generating a whole list just to get one value seems like a bad performance move(if the list gets big enough).

```python
next(iter(dict.keys()))
```
?
This only works for the zero indexed list accesses.

**Comment by raddessi** [comment]
Yeah good call, i think most of these are zero indexed so i will update. Thanks!

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by raddessi** [comment]
OK I made that change anywhere on the codebase it seemed relevant except for the tests. I only saw exception handling in 2 locations and those were both updated from IndexError to StopIteration

**Comment by raddessi** [comment]
If this is approved it can deprecate https://github.com/saltstack/salt/pull/53433 as this fixes the same issue across all modules not just napalm beacons. I KNEW I wasn't going crazy.. at least not yet anyway. This felt like total deja vu

**Comment by raddessi** [comment]
Hmm.. https://github.com/saltstack/salt/issues/56232 brings up the point that this is also relevant for `.items()[:]` indexing as well.. I'm going to look for occurrences of those as well

**Comment by raddessi** [comment]
Not many occurrences of that actually, hopefully that last commit covers it

**Review by s0undt3ch — APPROVED** [review]

**Review by cmcmarrow — DISMISSED** [review]

**Comment by raddessi** [comment]
Is there any way we could get this in to 3000.1?

**Review by s0undt3ch — DISMISSED** [review]
Looks good

**Review by DmitryKuzmenko — APPROVED** [review]

**Comment by raddessi** [comment]
I've set up a key for signing future commits now, can we skip this error here or is there a way to sign past commits?

**Review by DmitryKuzmenko — DISMISSED** [review]
Perfect

**Comment by raddessi** [comment]
well that wasn't as effective as I had hoped, will revert and then resolve manually

**Review by raddessi — COMMENTED** [review]
I think it lgtm again, way to go! I've been waiting for this project to get blacked for a while, thanks all :)

**Comment by DmitryKuzmenko** [comment]
Black is passed. And I found nothing related in centos failures. Re-ran the tests for now. Will check deeply if failed again.

**Review by DmitryKuzmenko — DISMISSED** [review]

**Review by cmcmarrow — DISMISSED** [review]

**Comment by dwoz** [comment]
> Hmm, generating a whole list just to get one value seems like a bad performance move(if the list gets big enough).
> 
> ```python
> next(iter(dict.keys()))
> ```
> 
> ?
> This only works for the zero indexed list accesses.

I can see problems arising from this since it's non-deterministic on python < 3.7. What I mean is that since dictionaries don't preserver insertion order until py3.7 on any other python version grabbing the first key is going to be non-deterministic. Seems like there might be cases that this will matter and could cause flaky test failures. Looks like this is the case currently as the code stands today. I'd prefer we look at each of these cases individually and make sure they are properly tested and we're doing what we expect. Perhaps some of these need to be sorted before grabbing the first item.

**Comment by raddessi** [comment]
Gotcha.. is there anything i can do to push this along myself or is this waiting on internal review now?

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by raddessi** [comment]
Hmm.. wasn't this the case before this pr also? Since the only thing we're changing here is how we extract data from the already existing dicts. The creation of the dicts isn't being changed here right?

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by dwoz** [comment]
Yes the behavior would be the same before these changes. My only point is that if we're taking the time to fix these instances we should also take the time to validate the functionality isn't buggy as it stands today. Hopefully that makes sense.

Black also needs to be run on these changes.

**Comment by raddessi** [comment]
Ok i see. Is that work i should do then or is it waiting on an internal review? It will take me quite some time to do this i think as im not familiar with most of these modules even slightly. Could this pr be accepted as long as we opened another one for the validation as this seems like scope creep or is that not allowed here? We're stuck running 2019.2.0 until these changes get merged in and i really want to update.

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by raddessi** [comment]
Oh and im pretty sure i blacked thia the day after that sep was merged ill check again though

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by raddessi** [comment]
I just saw https://github.com/saltstack/salt/pull/55690 referenced and this seems related also, mind if I add that one here as well?

**Comment by DmitryKuzmenko** [comment]
@raddessi I've looked at the PR I think it's better to keep it separate. It's logically a little bit different.

**Comment by DmitryKuzmenko** [comment]
Also one thought about this one. As I think it's better to drop the `.keys()` call from places where we need a first key. Just keep the `next(iter(dict))`. The same result, less calls, less time.

**Comment by raddessi** [comment]
Ah, ok I will remove.

**Comment by raddessi** [comment]
Can you also comment on my question here?

**Comment by DmitryKuzmenko** [comment]
Let me a couple of minutes. I'm doing review I have some more things. After this I'll check your question since it's not a simple one-line.

File: salt/beacons/cert_info.py
Comment (DmitryKuzmenko) [inline salt/beacons/cert_info.py]: Please remove the unnecessary call to `keys()` in all occurrences similar to this.

File: salt/beacons/cert_info.py
Comment (DmitryKuzmenko) [inline salt/beacons/cert_info.py]: Call duplication. It's better to move this line upper and re-use `cert_path` value in the call above.

File: salt/modules/bamboohr.py
Comment (DmitryKuzmenko) [inline salt/modules/bamboohr.py:78]: `next(iter(item.values()))` does the same but 20% faster. Could you please update all the occurrences that use `items()` call?

File: salt/modules/capirca_acl.py
Comment (DmitryKuzmenko) [inline salt/modules/capirca_acl.py:560]: `filter_name, filter_config = next(iter(filter_.items()))`

File: salt/modules/capirca_acl.py
Comment (DmitryKuzmenko) [inline salt/modules/capirca_acl.py:574]: Also oneline using `items`

File: salt/modules/highstate_doc.py
Comment (DmitryKuzmenko) [inline salt/modules/highstate_doc.py:659]: `*next(iter(w.values()))`
the same is below here

File: salt/pillar/vmware_pillar.py
Comment (DmitryKuzmenko) [inline salt/pillar/vmware_pillar.py]: The same value is calculated 3 times at once. Please put it to a variable.

File: salt/runners/net.py
Comment (DmitryKuzmenko) [inline salt/runners/net.py]: The call to `keys` is redundant here. `list.exted()` works good with dicts using the only keys.

**Review by raddessi — COMMENTED** [review]
lgtm now

**Comment by raddessi** [comment]
Am I meant to chmod these files? I haven't changed their modes but I do get errors from nox-py2 that I have to skip before commit:


```
~/d/salt ❯❯❯ gc                                                                                          master.py3-beacons ⬆ ✚ ◼
Linux Py3.5 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py3.5 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py3.5 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py3.5 Requirements.............................(no files to check)Skipped
Docs Py3.5 Requirements..............................(no files to check)Skipped
Linux Py3.5 Crypto Requirements......................(no files to check)Skipped
Darwin Py3.5 Crypto Requirements.....................(no files to check)Skipped
Windows Py3.5 Crypto Requirements....................(no files to check)Skipped
Lint Py3.5 Requirements..............................(no files to check)Skipped
Linux Py3.6 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py3.6 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py3.6 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py3.6 Requirements.............................(no files to check)Skipped
Docs Py3.6 Requirements..............................(no files to check)Skipped
Linux Py3.6 Crypto Requirements......................(no files to check)Skipped
Darwin Py3.6 Crypto Requirements.....................(no files to check)Skipped
Windows Py3.6 Crypto Requirements....................(no files to check)Skipped
Lint Py3.6 Requirements..............................(no files to check)Skipped
Linux Py3.7 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py3.7 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py3.7 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py3.7 Requirements.............................(no files to check)Skipped
Docs Py3.7 Requirements..............................(no files to check)Skipped
Linux Py3.7 Crypto Requirements......................(no files to check)Skipped
Darwin Py3.7 Crypto Requirements.....................(no files to check)Skipped
Windows Py3.7 Crypto Requirements....................(no files to check)Skipped
Lint Py3.7 Requirements..............................(no files to check)Skipped
isort....................................................................Passed
black....................................................................Passed
Lint Salt................................................................Failed
- hook id: nox-py2
- exit code: 1

Running session lint-salt-pre-commit
pip install --progress-bar=off distro
distro -j
Distro information:
{'codename': '',
 'id': 'fedora',
 'like': '',
 'version': '29',
 'version_parts': {'build_number': '', 'major': '29', 'minor': ''}}
Processing requirements/static/linux.in
pip install --progress-bar=off -r requirements/base.txt --constraint requirements/static/py3.7/linux.txt --constraint requirements/static/py3.7/linux-crypto.txt
pip install --progress-bar=off -r requirements/zeromq.txt --constraint requirements/static/py3.7/linux.txt --constraint requirements/static/py3.7/linux-crypto.txt
pip install --progress-bar=off -r requirements/pytest.txt --constraint requirements/static/py3.7/linux.txt --constraint requirements/static/py3.7/linux-crypto.txt
pip install --progress-bar=off -r requirements/static/linux.in --constraint requirements/static/py3.7/linux.txt --constraint requirements/static/py3.7/linux-crypto.txt
pip install --progress-bar=off -r requirements/static/lint.in --constraint requirements/static/py3.7/lint.txt
pylint --rcfile=.pylintrc --disable=I salt/states/mdadm_raid.py salt/modules/win_lgpo.py salt/modules/postgres.py salt/modules/zabbix.py salt/modules/rabbitmq.py salt/states/mount.py salt/beacons/napalm_beacon.py salt/modules/capirca_acl.py salt/cli/batch.py salt/modules/opkg.py salt/runners/virt.py salt/states/dellchassis.py salt/pillar/vmware_pillar.py salt/modules/acme.py salt/beacons/cert_info.py salt/utils/dns.py
************* Module salt.states.mdadm_raid
salt/states/mdadm_raid.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
/home/raddessi/.cache/pre-commit/repoz59pzhke/py_env-python3/lib/python3.7/site-packages/isort/settings.py:295: UserWarning: Found /home/raddessi/dev/salt/pyproject.toml with [tool.isort] section, but toml package is not installed. To configure isort with /home/raddessi/dev/salt/pyproject.toml, install with 'isort[pyproject]'.
  file_path))
************* Module salt.modules.win_lgpo
salt/modules/win_lgpo.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.modules.postgres
salt/modules/postgres.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.modules.zabbix
salt/modules/zabbix.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.modules.rabbitmq
salt/modules/rabbitmq.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.states.mount
salt/states/mount.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.beacons.napalm_beacon
salt/beacons/napalm_beacon.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.modules.capirca_acl
salt/modules/capirca_acl.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.cli.batch
salt/cli/batch.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.modules.opkg
salt/modules/opkg.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.runners.virt
salt/runners/virt.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.states.dellchassis
salt/states/dellchassis.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.pillar.vmware_pillar
salt/pillar/vmware_pillar.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.modules.acme
salt/modules/acme.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.beacons.cert_info
salt/beacons/cert_info.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664
************* Module salt.utils.dns
salt/utils/dns.py:1: [E0599(file-perms), ] Module file has the wrong file permissions(expected 0o644): 0o664

-----------------------------------
Your code has been rated at 9.90/10

Command pylint --rcfile=.pylintrc --disable=I salt/states/mdadm_raid.py salt/modules/win_lgpo.py salt/modules/postgres.py salt/modules/zabbix.py salt/modules/rabbitmq.py salt/states/mount.py salt/beacons/napalm_beacon.py salt/modules/capirca_acl.py salt/cli/batch.py salt/modules/opkg.py salt/runners/virt.py salt/states/dellchassis.py salt/pillar/vmware_pillar.py salt/modules/acme.py salt/beacons/cert_info.py salt/utils/dns.py failed with exit code 2
Session lint-salt-pre-commit failed.

Lint Tests...........................................(no files to check)Skipped
~/d/salt ❯❯❯ 
```

File: salt/utils/dns.py
Comment (DmitryKuzmenko) [inline salt/utils/dns.py:98]: 1. Py3 `dict_values` has no `index` method.
2. This looks weird at all. I know it's not your code but...
Please change it to something like `return {v: k for k, v in ref.items()}[lookup]

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]

**Comment by DmitryKuzmenko** [comment]
@raddessi oh, you're faster than me =))
I've done the promised review. Checking your updates.

**Comment by raddessi** [comment]
Oh im seeing other comments that you added but i didnt address.. huh ill have to see why i didnt see them before. Ill make the other changes and update, sorry for the false flag.

**Comment by DmitryKuzmenko** [comment]
Don't excuse, you're good! I've just added this review. It took time.
About your nox-py2 errors they look strange. How do you run pre-commit? How did you install that?

**Comment by raddessi** [comment]
For precommit i just did a normal `pre-commit install` to add the hook

Sent from ProtonMail mobile

-------- Original Message --------

File: salt/beacons/cert_info.py
Comment (raddessi) [reply salt/beacons/cert_info.py]: Its duplicated but if we move it above then the indexed get of `cert_path` on line 103 would be on the next item, not the current. I think in it's current state this is probably supposed to be:
```

            try:
                notify_days = cert_path[next(iter(cert_path))].get(
                    "notify_days", global_notify_days
                )
                
            except StopIteration as exc:
                log.error("Unable to load certificate %s (%s)", cert_path, exc)
                continue
            else:
                cert_path = next(iter(cert_path))
```
So I'm going to update with:
```
            try:
                next_cert_path = next(iter(cert_path))
                notify_days = cert_path[next_cert_path].get(
                    "notify_days", global_notify_days
                )
            except StopIteration as exc:
                log.error("Unable to load certificate %s (%s)", cert_path, exc)
                continue
            else:
                cert_path = next_cert_path
```
if that seems ok

Comment (raddessi) [reply salt/beacons/cert_info.py]: Actually, that's probably more correct as
```
            try:
                next_cert_path = next(iter(cert_path))
            except StopIteration as exc:
                log.error("Unable to load certificate %s (%s)", cert_path, exc)
                continue
            else:
                notify_days = cert_path[next_cert_path].get(
                    "notify_days", global_notify_days
                )
                cert_path = next_cert_path
```

File: salt/modules/bamboohr.py
Comment (raddessi) [reply salt/modules/bamboohr.py:78]: Did you mean `next(iter(item.items()))` in this case?

File: salt/modules/capirca_acl.py
Comment (raddessi) [reply salt/modules/capirca_acl.py:560]: haha I should have seen that. Thanks, will fix

File: salt/modules/capirca_acl.py
Comment (raddessi) [reply salt/modules/capirca_acl.py:574]: done

File: salt/pillar/vmware_pillar.py
Comment (raddessi) [reply salt/pillar/vmware_pillar.py]: done

File: salt/modules/highstate_doc.py
Comment (raddessi) [reply salt/modules/highstate_doc.py:659]: done

File: salt/utils/dns.py
Comment (raddessi) [reply salt/utils/dns.py:98]: Oh wow is that what this is doing.. lol I didn't spend enough time to try and figure it out. Done

File: salt/beacons/cert_info.py
Comment (raddessi) [reply salt/beacons/cert_info.py]: I think these should all be done, like you said we just updated at the same time.

File: salt/runners/net.py
Comment (raddessi) [reply salt/runners/net.py]: Ah this is from the other PR I included then removed.

File: salt/modules/highstate_doc.py
Comment (raddessi) [inline salt/modules/highstate_doc.py:659]: `_format_markdown_requisite(next_w_item[0], next_w_item[1],)` is actually called itself up to 4 times, should that instead be the thing moved up?

**Review by raddessi — COMMENTED** [review]
one more possible change

**Comment by raddessi** [comment]
@DmitryKuzmenko how does this look now?

File: salt/modules/highstate_doc.py
Comment (DmitryKuzmenko) [reply salt/modules/highstate_doc.py:659]: I see there are 4 almost equal cases that looks like could be shrunken to one case.
Everything We could do something like this here:
```python
for comment, key in (("run or update after changes in:\n", "watch"), ("after changes, run or update:\n", "watch_in"), ("require:\n", "require"), ("required in:\n", "require_in")):
    reqs = s.get(key, [])
    if reqs:
        requisites += comment
        for w in reqs:
            requisites += _format_markdown_requisite(*next(iter(w.items())))
```

File: salt/modules/postgres.py
Comment (DmitryKuzmenko) [inline salt/modules/postgres.py:896]:
```python
'ALTER TABLESPACE "{0}" SET ({1} = {2})'.format(name, *(next(iter(items()))))
```

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]
I'm very sorry but here's another one bunch of change requests.

**Comment by bryceml** [comment]
re-run all full

**Comment by raddessi** [comment]
No need to apologize, the comments look good. I'll try to get this up to date tomorrow. Thanks!

File: salt/beacons/napalm_beacon.py
Comment (raddessi) [reply salt/beacons/napalm_beacon.py:299]: done

File: salt/beacons/napalm_beacon.py
Comment (raddessi) [reply salt/beacons/napalm_beacon.py:325]: done

File: salt/modules/bamboohr.py
Comment (raddessi) [reply salt/modules/bamboohr.py:80]: nice! I would not have thought of this. Done for both

File: salt/modules/bamboohr.py
Comment (raddessi) [reply salt/modules/bamboohr.py:149]: done

File: salt/modules/capirca_acl.py
Comment (raddessi) [reply salt/modules/capirca_acl.py:417]: done

File: salt/modules/postgres.py
Comment (raddessi) [reply salt/modules/postgres.py:896]: much better, done

**Comment by raddessi** [comment]
Good call, I completely missed one of the loops before. Latest requested changes should be in now @DmitryKuzmenko

File: salt/modules/bamboohr.py
Comment (raddessi) [reply salt/modules/bamboohr.py]: great, done

File: salt/modules/highstate_doc.py
Comment (raddessi) [reply salt/modules/highstate_doc.py]: Ugh how did I not see that. Thanks :) done

**Review by DmitryKuzmenko — APPROVED** [review]

**Comment by DmitryKuzmenko** [comment]
Perfect! Thank you!

**Comment by dwoz** [comment]
Has anyone validated there is test coverage for the methods changing in this PR?

**Comment by raddessi** [comment]
@dwoz not yet. Last I spoke with @DmitryKuzmenko we were both going to try to get that done so this could get in. I have not had any spare time yet unfortunately.

**Comment by dwoz** [comment]
Merging this since @DmitryKuzmenko has no easy way of verifying the test coverage. For the record I'm not a huge fan of the precedence this sets.

**Pull request merged**
